timemachine/cpp/src/kernels/k_centroid_restraint.cuh

```diff
@@ -2,88 +2,82 @@
 
 #include "k_fixed_point.cuh"
 
-__device__ const double PI = 3.14159265358979323846;
+template<typename RealType>
+void __global__ k_calc_centroid(
+    const double * __restrict__ d_coords,  // [n, 3]
+    const int * __restrict__ d_group_a_idxs,
+    const int * __restrict__ d_group_b_idxs,
+    const int N_A,
+    const int N_B,
+    RealType * d_centroid_a, // [3]
+    RealType * d_centroid_b // [3]
+    ){
+    int t_idx = blockDim.x*blockIdx.x + threadIdx.x;
+    if (N_A + N_B <= t_idx) {
+        return;
+    }
+    int group_idx = t_idx < N_A ? t_idx : t_idx - N_A;
+    RealType * centroid = t_idx < N_A ? d_centroid_a : d_centroid_b;
+    const int * cur_array = t_idx < N_A ? d_group_a_idxs : d_group_b_idxs;
+    for(int d=0; d < 3; d++) {
+        atomicAdd(centroid + d, d_coords[cur_array[group_idx]*3+d]);
+    }
+ }
 
 template<typename RealType>
 void __global__ k_centroid_restraint(
-    const int N,     // number of bonds
-    const double *coords,  // [n, 3]
-    const int *group_a_idxs,
-    const int *group_b_idxs,
+    // const int N,     // number of bonds, ignore for now
+    const double * __restrict__ d_coords,  // [n, 3]
+    const int * __restrict__ d_group_a_idxs,
+    const int * __restrict__ d_group_b_idxs,
     const int N_A,
     const int N_B,
-    // const double *masses, // ignore masses for now
+    const RealType * __restrict__ d_centroid_a, // [3]
+    const RealType * __restrict__ d_centroid_b, // [3]
+    // const double *d_masses, // ignore d_masses for now
     const double kb,
     const double b0,
-    unsigned long long *grad_coords,
-    unsigned long long *energy) {
-
-    // (ytz): ultra shitty inefficient algorithm, optimize later
-    const auto t_idx = blockDim.x*blockIdx.x + threadIdx.x;
+    unsigned long long *d_du_dx,
+    unsigned long long *d_u) {
 
-    if(t_idx != 0) {
+    if (!d_u && !d_du_dx) {
```

Comment:
Unless this is for performance reasons, the code should naturally be able to handle this (and it does later below). This is generally unwieldy for more complicated kernels (which may also request du_dl/du_dp) and the combinations blow up fast.